lib/sapi.py

```diff
@@ -236,8 +237,25 @@ def redirect(self, url):
     self.write_text(redirect_notice(url))
 
   def getenv(self, name, value=None):
-    ret = os.environ.get(name, value)
-    if self._iis and name == 'PATH_INFO' and ret:
+    # we should always use UTF-8 for environment variable.
+    if sys.getfilesystemencoding() == 'UTF-8':
+      ret = os.environ.get(name, value)
+    else:
+      if os.supports_bytes_environ:
+        if isinstance(value, str):
+          value = value.encode('utf-8', 'surrogateescape')
+        ret = os.environb.get(name.encode('utf-8'), value)
+      else:
+        ret = os.environ.get(name, value)
+        if isinstance(ret, str):
+          ret = ret.encode(sys.getfilesystemencoding(), 'surrogateescape')
+      if isinstance(ret, bytes):
+        ret = ret.decode('utf-8','surrogateescape')
+    if name in ('PATH_INFO', 'SCRIPT_NAME') and ret:
+      # Also, we should interpret URL encoding as UTF-8 stream, however,
+      # we should hold whole sequence even if it causes error.
+      ret = _unquote(ret, 'utf-8', 'surrogateescape')
```

Comment:
I'm pretty sure this is at odds both with the CGI RFC (https://tools.ietf.org/html/rfc3875#section-4.1.5 ; https://tools.ietf.org/html/rfc3875#section-4.1.13) and with common practice under the likes of nginx and Apache HTTP Server.  `PATH_INFO` and `SCRIPT_NAME` are _not_ URI-encoded by the gateway, thus we should not be trying to use `unquote()` on them.

Reply:
I see, I confirmed you are right.  I'll fix it and revert 4757ef9.

Reply:
> I see, I confirmed you are right. I'll fix it and revert [4757ef9](https://github.com/viewvc/viewvc/commit/4757ef995998458dd013c5cf305c395a8a3cacea).

done.